[pt][quant] Adding quantized clamp kernel

dskhudia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30541 [pt][quant] Adding quantized clamp kernel**

Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

dskhudia committed:
[pt][quant] Adding quantized clamp kernel



Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -0,0 +1,37 @@
+static auto registry = c10::RegisterOperators().op(
+    "quantized::clamp(Tensor qx, Scalar min, Scalar max) -> Tensor qy",
+    c10::RegisterOperators::options().kernel<QClamp>(
+        TensorTypeId::QuantizedCPUTensorId));
```
I think we actually want to switch to registering new ops via `native_functions.yaml`, however someone needs to add support for namespaces there (i.e `quantized::`), so this is OK for now

dskhudia replied:
Yes. The corresponding float operation is torch.clamp. So eventually this would be torch.quantized.clamp?

jamesr66a reviewed (changes requested):
Cool! Unfortunately this is going to need yet more specializations of minimum for the <AVX fallback, i.e. at vec256_qint.h lines 778, 848, and 853. Otherwise looks good

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

jamesr66a reviewed (approved):
Great!

dskhudia commented:
The added test is failing. I am looking into it.

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -71,6 +72,13 @@ def cat(self, x, dim=0):
+        r = torch.clamp(x, min_, max_)
+        r = self.activation_post_process(r)
+        return r
+
```
we probably don't need this if clamp is using scale/zero_point from input

dskhudia replied:
Let me remove this.

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 36360fb:
* **2/3** broken upstream at merge base 5b03ff0 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=5b03ff0a09d43d721067e39da10aa23edc6997cd&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **1/3** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=36360fb7c22104f49394cf4d9f01e03577fdbbfa).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3910451) pytorch_windows_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72644253))
```
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(192): error C2143: syntax error: missing ';' before '<'
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(192): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(193): error C2653: 'DeviceType': is not a class or namespace name
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(205): error C2061: syntax error: identifier 'DeviceType'
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(211): error C2065: 'DeviceType': undeclared identifier
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(211): error C2146: syntax error: missing ')' before identifier 'type'
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(211): error C2143: syntax error: missing ';' before '{'
C:/Users/circleci/project\c10/core/impl/DeviceGuardImplInterface.h(211): fatal error C1003: error count exceeds 100; stopping compilation
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_swish_op.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_swish_op.cu.obj
CMake Error at torch_cuda_generated_swish_op.cu.obj.Release.cmake:281 (message):
  Error generating file
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_swish_op.cu.obj


[2207/2731] cmd.exe /C "cd /D C:\Users\circleci\project\build\caffe2\CMakeFiles\torch_cuda.dir\operators && "C:\Program Files\CMake\bin\cmake.exe" -E make_directory C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/. && "C:\Program Files\CMake\bin\cmake.exe" -D verbose:BOOL=ON -D build_configuration:STRING=Release -D generated_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_thresholded_relu_op.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_thresholded_relu_op.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/torch_cuda_generated_thresholded_relu_op.cu.obj.Release.cmake"
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_thresholded_relu_op.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/./torch_cuda_generated_thresholded_relu_op.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/torch_cuda_generated_thresholded_relu_op.cu.obj.NVCC-depend
C:/Users/circleci/project/build/win_tmp/bin/nvcc -M -D__CUDACC__ C:/Users/circleci/project/caffe2/operators/thresholded_relu_op.cu -o C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/operators/torch_cuda_generated_thresholded_relu_op.cu.obj.NVCC-depend -ccbin cl.exe -m64 -Dtorch_cuda_EXPORTS -DTORCH_CUDA_BUILD_MAIN_LIB -DTH_BLAS_MKL -D_OPENMP_NOFORCE_MANIFEST -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -D_CRT_SECURE_NO_DEPRECATE=1 -DWIN32_LEAN_AND_MEAN -DMAGMA_V2 -DAT_PARALLEL_OPENMP=1 -Xcompiler ,\"/DWIN32\",\"/D_WINDOWS\",\"/GR\",\"/w\",\"/MP\",\"/bigobj\",\"-openmp:experimental\",\"-DHAVE_AVX_CPU_DEFINITION\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"/MD\",\"/O2\",\"/Ob2\",\"/w\",\"/MP\",\"/bigobj\" -Xcompiler /w -w -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_75,code=sm_75 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -Xcompiler -MD --expt-relaxed-constexpr --expt-extended-lambda -Xcompiler /wd4819 -Xcompiler /wd4503 -Xcompiler /wd4190 -Xcompiler /wd4244 -Xcompiler /wd4251 -Xcompiler /wd4275 -Xcompiler /wd4522 -Wno-deprecated-gpu-targets --expt-extended-lambda -gencode arch=compute_75,code=sm_75 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC "-IC:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include" -IC:/Users/circleci/project/build/aten/src -IC:/Users/circleci/project/aten/src -IC:/Users/circleci/project/build -IC:/Users/circleci/project -IC:/Users/circleci/project/cmake/../third_party/googletest/googlemock/include -IC:/Users/circleci/project/cmake/../third_party/googletest/googletest/include -IC:/Users/circleci/project/third_party/protobuf/src -IC:/Users/circleci/project/build/win_tmp/mkl/include -IC:/Users/circleci/project/cmake/../third_party/benchmark/include -IC:/Users/circleci/project/cmake/../third_party/eigen -IC:/Jenkins/Miniconda3/include -IC:/Jenkins/Miniconda3/lib/site-packages/numpy/core/include -IC:/Users/circleci/project/cmake/../third_party/pybind11/include -I/opt/rocm/hip/include -I/include -IC:/Users/circleci/project/cmake/../third_party/cub -IC:/Users/circleci/project/build/caffe2/contrib/aten -IC:/Users/circleci/project/third_party/onnx -IC:/Users/circleci/project/build/third_party/onnx -IC:/Users/circleci/project/third_party/foxi -IC:/Users/circleci/project/build/third_party/foxi -IC:/Users/circleci/project/build/win_tmp/magma/include "-IC:/Program Files/NVIDIA Corporation/NvToolsExt/include" -IC:/Users/circleci/project/build/caffe2/aten/src/TH -IC:/Users/circleci/project/aten/src/TH -IC:/Users/circleci/project/build/caffe2/aten/src/THC -IC:/Users/circleci/project/aten/src/THC -IC:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
thresholded_relu_op.cu
```
### 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_6_gcc5_4_test](https://dr.pytorch.org/build-details.html?build_id=72675950)
* [pytorch_xla_linux_xenial_py3_6_clang7_test](https://dr.pytorch.org/build-details.html?build_id=72677826)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

dskhudia committed:
Update on "[pt][quant] Adding quantized clamp kernel"


Differential Revision: [D18739628](https://our.internmc.facebook.com/intern/diff/D18739628/)

[ghstack-poisoned]
